### PR TITLE
feat: add solar pro llm and document parser

### PR DESCRIPTION
**Pull request by JuHyung-Son**
# Description

Add new llm model and readers api of upstage.

New llm model, `solar-pro` can be found here https://developers.upstage.ai/docs/apis/chat .
New document parse model `Document Parse` can be found here https://developers.upstage.ai/docs/apis/document-parse .

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [ ] Yes
- [x] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [x] Yes
- [ ] No

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [ ] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-integrations/llms/llama-index-llms-upstage/llama_index/llms/upstage/base.py [hunk 1]
```diff
@@ -207,3 +231,50 @@ def get_num_tokens_from_message(self, messages: Sequence[ChatMessage]) -> int:
                 )
         num_tokens += tokens_suffix
         return num_tokens
+
+    @llm_retry_decorator
+    def _chat(self, messages: Sequence[ChatMessage], **kwargs: Any) -> ChatResponse:
```
Comment (logan-markewich) [inline llama-index-integrations/llms/llama-index-llms-upstage/llama_index/llms/upstage/base.py:236]: curious why we subclass _chat, but not any other method like _stream_chat or _achat, etc.

Comment (logan-markewich) [reply llama-index-integrations/llms/llama-index-llms-upstage/llama_index/llms/upstage/base.py:236]: I guess we only support parse documents with `.chat()` ?

File: llama-index-integrations/llms/llama-index-llms-upstage/llama_index/llms/upstage/base.py [hunk 2]
```diff
@@ -207,3 +231,50 @@ def get_num_tokens_from_message(self, messages: Sequence[ChatMessage]) -> int:
...
+
+        if isinstance(file_path, list):
+            file_titles = [os.path.basename(path) for path in file_path]
+        else:
+            file_titles = [os.path.basename(file_path)]
+
+        for i, doc in enumerate(docs):
+            file_title = file_titles[min(i, len(file_titles) - 1)]
+            document_contents += f"{file_title}:\n{doc.text}\n\n"
+        print("DOCUMENT CONTENTS", document_contents)
```
Comment (logan-markewich) [inline llama-index-integrations/llms/llama-index-llms-upstage/llama_index/llms/upstage/base.py:279]: lets remove this print

File: llama-index-integrations/llms/llama-index-llms-upstage/pyproject.toml [hunk 3]
```diff
@@ -35,9 +35,9 @@ version = "0.2.1"
 [tool.poetry.dependencies]
 python = ">=3.8.1,<4.0"
 openai = "^1.21.2"
-llama-index-llms-openai = "^0.2.0"
 tokenizers = "^0.19.1"
 llama-index-core = "^0.11.0"
+llama-index-llms-openai = "^0.2.0"
```
Comment (logan-markewich) [inline llama-index-integrations/llms/llama-index-llms-upstage/pyproject.toml:40]: we need to add the upstage reader as a dependency now right?

Comment (logan-markewich) [reply llama-index-integrations/llms/llama-index-llms-upstage/pyproject.toml:40]: (should also bump the version of the integration package here)

File: llama-index-integrations/llms/llama-index-llms-upstage/poetry.lock
Comment (logan-markewich) [inline llama-index-integrations/llms/llama-index-llms-upstage/poetry.lock:1]: no need to check this in

File: llama-index-integrations/readers/llama-index-readers-upstage/poetry.lock
Comment (logan-markewich) [inline llama-index-integrations/readers/llama-index-readers-upstage/poetry.lock:1]: no need to check this in

**Review by logan-markewich — COMMENTED** [review]
Looks good, just a few things to clean up

File: llama-index-integrations/llms/llama-index-llms-upstage/pyproject.toml (hunk 3 above)
Comment (logan-markewich) [reply llama-index-integrations/llms/llama-index-llms-upstage/pyproject.toml:40]: Or, maybe we need to publish the reader first actually

Comment (JuHyung-Son) [reply llama-index-integrations/llms/llama-index-llms-upstage/pyproject.toml:40]: Yeah you are right. we should publish reader first.
> we need to add the upstage reader as a dependency now right?

yes, we need to add it

File: llama-index-integrations/llms/llama-index-llms-upstage/llama_index/llms/upstage/base.py (hunk 1 above)
Comment (JuHyung-Son) [reply llama-index-integrations/llms/llama-index-llms-upstage/llama_index/llms/upstage/base.py:239]: There was a mistake in the llama index interface. Fixed.

File: llama-index-integrations/readers/llama-index-readers-upstage/llama_index/readers/upstage/document_parse.py [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+    """
+
+    def __init__(
+        self,
+        api_key: Optional[str] = None,
+        base_url: str = DOCUMENT_PARSE_BASE_URL,
+        model: str = DOCUMENT_PARSE_DEFAULT_MODEL,
+        split: SplitType = "none",
+        ocr: OCR = "auto",
+        output_format: OutputFormat = "html",
```
Comment (cyzhzhd) [inline llama-index-integrations/readers/llama-index-readers-upstage/llama_index/readers/upstage/document_parse.py:122]: This should be List[OutputFormat]

**Pull request merged**